galaxyproject/galaxy
# [23.1] Enable job resubmissions in k8s runner

This PR brings working resubmissions to the k8s runner (largely inspired in the LSF cli runner where I have used resubmissions for many years), plus a few additional changes:

- [x] Correctly detects OOM and labels jobs.
- [x] For the above to happen, the k8s cluster and the runner had to be made aware again that the job had failed. For some years, the k8s runner was made to detect every single job as a success, and then some other Galaxy process outside decided if this was actually a successful or failed job.
- [x] Introduces the usage of job_spec_backoff_limit, so that the runner doesn't blindly resubmit directly any jobs and that can be delegated with more confidence to the Galaxy resubmission handler. In time, the pod retrials variable and its usage should be deprecated, to simplify the runner's logic.
- [x] corrects the empty stdout and stderr of failed jobs. Currently, failed jobs that are detected by the runner and not the external galaxy setup are missing stdout and stderr. This should be somehow consumed from the job files for the tool stderr and tool stdout.
- [x]  Move to pykube-ng
- [x] Change ingress versions and modify creation dictionaries to fit newer version.
- [x] Test that interactive tools work correctly.
- [x] Interactive tools also work with resubmissions.

This solution both avoids extra API calls to the k8s masters (desirable in scenarios with many jobs) and reduces any chances of extra pods being sent to run.

Takes care of #15118

## How to test the changes?
(Select all options that apply)

Please let me know where I should add testing for the resubmission capability, and what I can reuse from tests of this capability in other runners.

## License
